src/Components/Components/src/NavigationManager.cs

```diff
@@ -274,6 +282,87 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
         }
     }
 
+    /// <summary>
+    /// Notifies the registered handlers of the current location change.
+    /// </summary>
+    /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI.</param>
+    /// <param name="isNavigationIntercepted">Whether this navigation was intercepted from a link.</param>
+    /// <param name="continueNavigation">An action that continues the navigation based on the provided <see cref="LocationChangingResult"/>.</param>
+    protected void NotifyLocationChanging(string uri, bool isNavigationIntercepted, Action<LocationChangingResult> continueNavigation)
+    {
+        _locationChangingCts?.Cancel();
+        _locationChangingCts = null;
+
+        if (_locationChangingHandlers.Count == 0)
+        {
+            continueNavigation(new(canceled: false, exception: null));
+            return;
+        }
+
+        _locationChangingCts = new();
+
+        var context = new LocationChangingContext(uri, isNavigationIntercepted, _locationChangingCts);
+        var cancellationToken = _locationChangingCts.Token;
+        var handlerCount = _locationChangingHandlers.Count;
+        var locationChangingHandlersCopy = ArrayPool<LocationChangingHandler>.Shared.Rent(handlerCount);
+        var locationChangingTasks = ArrayPool<Task>.Shared.Rent(handlerCount);
+
+        _locationChangingHandlers.CopyTo(locationChangingHandlersCopy);
+
+        for (var i = 0; i < handlerCount; i++)
+        {
+            locationChangingTasks[i] = locationChangingHandlersCopy[i].Invoke(context);
+        }
+
+        Task.WhenAll(locationChangingTasks.Take(handlerCount)).WaitAsync(cancellationToken).ContinueWith(t =>
+        {
+            continueNavigation(new(cancellationToken.IsCancellationRequested, t.Exception));
+        }, TaskScheduler.Current);
```

Comment:
@MackinnonBuck this can be problematic. `ContinueWith` != `await task`, for example the execution context is not flowed via `ContinueWith`. 

I would suggest trying to always rely on async/await when possible and only skip doing that if you have a really good reason to do so.

In this case, you could move the code to a local helper method and just call it from here.

Reply:
> The key is that we don't await the task on the hub, we will just "deliver" the locationchanging event.

How do we avoid awaiting? If there are multiple handlers, something has to `WhenAll` them, doesn't it?